### PR TITLE
fix: prevent duplicated slugs for headings with the same title

### LINKED ISSUE #154: Rename the slugs of duplicated headings
**Is your feature request related to a problem? Please describe.**

When using `@islands/headings`, if multiple headings have the same name, now they will share the same slug, and then the anchor will only work for the first heading with the shared name.

**Describe the solution you'd like**

Rename the slug when multiple headings have the same name. For example, if the slug `foo` is already used, we can give `foo-2` to the next heading whose name is `foo`, and then give `foo-3` to the next one whose name is `foo`, and then maybe give `foo-3-2` to a heading whose name is `foo-3`.

**Describe alternatives you've considered**

Avoid duplicated headings when writing Markdown.

### PR DESCRIPTION
**Pull request by ouuan**
This resolves #154.

A counter is used to track the number of duplicated titles.

To pass the counter to the slug generator, the API of `generateSlug` is modified.

ids are taken into account when generating distinct slugs, but duplicated ids are not resolved and kept as-is, because it's the user's responsibility to not set duplicated ids.

An example (test suite):

    title     id      slug
    foo       none    foo
    bar       none    bar-2
    foo       none    foo-2
    foo       bar     bar
    foo-2     none    foo-2-2
    foo-4     none    foo-4
    foo       none    foo-3
    foo       none    foo-5
    bar       none    bar-3
    foo-2     none    foo-2-3
    foo-2-2   none    foo-2-2-2
    foo       bar     bar

A consideration is that the type and initialization of `counter` limit the freedom of `HeadingOptions.slug`. Maybe we can change the parameter to a generic `T` and add `HeadingOption.init: (ast: Parent) => T` or use `any`.

### CODE REVIEW

File: packages/headings/src/headings.ts [hunk 1]
```diff
@@ -66,12 +69,38 @@ export const rehypePlugin: HeadingPlugin = ({ slug = generateSlug } = {}) => (as
...
+function initCounter (ast: any) {
+  const { children } = ast as Parent
+  const counter = new Map<string, number>()
+  // add id in the counter first because when id is set it is used as slug
+  children.forEach((child: any) => {
+    const id = child.properties?.id
+    if (id) counter.set(id, 1)
+  })
+  return counter
+}
```
Comment (ElMassimo) [inline packages/headings/src/headings.ts:81]: Let's skip this step, as it's not recursive, so ids in nested elements won't be detected, and this check doesn't prevent that an id outside the document (in a layout) conflicts with the headings.

The overhead of walking the entire document recursively is not acceptable for this plugin (though you can create a separate rehype plugin if you need to achieve that).

**Review by ElMassimo — APPROVED** [review]
Looks great, thanks for this Yusuf! 😃 

Would like to remove `initData`, for the [reasons mentioned here](https://github.com/ElMassimo/iles/pull/155/files#r920602664).

**Comment by ElMassimo** [comment]
Thanks! I'll make a release later this week 😃

**Comment by ouuan** [comment]
Just a reminder that this is still not released.

**Comment by ElMassimo** [comment]
@ouuan Must have missed it last time, released in `@islands/headings@0.8.1`, thanks!

**Pull request merged**